[pull] primary from flybywiresim:primary (#469)
Valastiri/fbw-userguide: docs/fbw-a32nx/feature-guides/local-api-server.md

![Local API Server](../assets/local-api-server/local-api-server.png "Local API Server"){loading=lazy}

Double click on the `local-server.exe` file to start the Local API Server. The Local API Server does not need any command line paramters and is configured completely by the configuration file. See [Local API Server Configuration](#local-api-server-configuration)

**You will need to keep this window open and the Local API Server running to use the MCDU Web Interface**.

!!! info "Local API Server Command Window"
    ![Local API Server Command Window](../assets/local-api-server/local-api-server-cmd-window.png "Local API Server Command Window"){loading=lazy}

## Local API Server Configuration

The Local API configuration file can be found here:

!!! note "Local API Server Configuration File"
    ![Local API Server Configuration File](../assets/local-api-server/configuration-properties-file.png "configuration-properties-file")

```json title="Local API Server Configuration"
{
	"server": {
		"port": 8380
	},
	"printer": {
		"enabled": false,
		"printerName": null,
		"fontSize": 19,
		"paperSize": "A4",
		"margin": 30
	}
}
```

> 

Reviewer: We should try to keep this docs page updated then

This is typically the case if both, the PC and the device use the same Internet router and the same WiFi.

To check this you can look at the IP address the MCDU server has listed in its command window while starting.

![MCDU Server Command Window](../assets/local-api-server/local-api-server-cmd-window.png "MCDU Server Command Window"){loading=lazy}

Or check it as described here: [Find your IP address in Windows](https://support.microsoft.com/en-au/windows/find-your-ip-address-in-windows-f21a9bbc-c582-55cd-35e0-73431160a1b9)

In the example above the IP address is `192.168.1.19`.

Your device should typically also have an IP address starting with `192.168.1.x` in this case.

You can check this in the device's network configuration.

??? tip "Examples for Device Network Information (click to expand)"
    **iPad:**

    ![iPad Network Info](../assets/mcdu-server/ipad-network-info.png "iPad Network Info"){loading=lazy}

    **Samsung Android:**

    ![Samsung Network Info](../assets/mcdu-server/samsung-network-info.png "Samsung Network Info"){loading=lazy}

> If you still can't connect to the MCDU server your firewall might block the traffic. See next chapter.

Reviewer 2: ```suggestion
If you still can't connect to the MCDU server your firewall might be blocking the traffic. See next chapter.
```

## Firewall Configuration

!!! danger "Disclaimer"

>     Changing Windows settings, especially security settings like the firewall comes with certain risks. Please do not change these settings if you are not comfortable doing so. FlyByWire Simulations cannot take any responsibility for any issues caused by your changes to Windows or security settings.

Reviewer 2: ```suggestion
    Changing Windows settings, especially security settings like the firewall comes with certain risks. Please do not change these settings if you are not comfortable doing so. FlyByWire Simulations does not take any responsibility for any issues caused by your changes to Windows or security settings.
```
Change the "voice" here so that it's less us not wanting to take responsibility and just out right don't (when config'in system critical things). Use this at will.

	<Disabled>False</Disabled>
	
	<!-- NEW APPLICATIONS TO START HERE -->
	
</SimBase.Document>
```

Add the following section before the `</SimBase.Document>` tag and change the `YOUR_COMMUNITY_FOLDER` placeholder with the path to your Community folder.

``` xml title="exe.xml (partial)"
	<Launch.Addon>
		<Name>FlyByWire Local API Server</Name>
		<Disabled>False</Disabled>
		<Path>YOUR_COMMUNITY_FOLDER\flybywire-aircraft-a320-neo\local-server.exe</Path>
	</Launch.Addon>
```
Now every time Microsoft Flight Simulator starts it will also start the Local API Server.

!!! warning ""
    Unfortunately MSFS does not stop the automatically started applications after shutting down and you need to close the Local API Server window manually.

    If you do not do this there will be an error message the next time you start MSFS or the Local API Server as it is already running.

## Logfile

The command window running the Local API Server displays the servers log file information and provides important information about the started services, the necessary urls and ip addresses.

It also shows the available printer to help with configuring the Remote MCDU service for local printing.

> The log is also stored in a file in this folder:

Reviewer 3: `The log file is also stored in a file in this folder, formatted by date`

    The Local API Server and MCDU Server are only available in the Development Version.

??? warning "Error: listen EADDRINUSE: address already in use xxxx"
    The web socket port you are using is already in use.

    Solution: [Port is Occupied](#port-is-occupied)

??? warning "Local API Server not connecting to simulator although flight is Started"
    The Local API Server does not connect to the flight simulator although the flight is started and aircraft is loaded.

    The port 8380 is already in use.

    Solution: [Port is Occupied](#port-is-occupied)

??? warning "Local API Server Error: Failed to load printers"

    If you you start the server and get this error messages:

    ``` cmd title="Windows Command Line"
    Error: Failed to load printers.
    Make sure the "Print Spooler" Windows service is running.
    ```
    Solution: Make sure the "Printer Spooler" Windows service is turned on:

    ![mcdu-windows-service-printer](../assets/mcdu-server/mcdu-windows-service-printer.png){loading=lazy}

    You can find the Services Manager by pressing ++windows+r++ and then type `services.msc`.

??? warning "Local API Server Error: `...Encoding not recognized...`"

>     When you start the server and and it crashes immediately with an error message mentioning `...Encoding not recognized...` you Windows most likely uses a rare UTF-8 based encoding.

Reviewer 2: ```suggestion
    When you start the server and and it crashes immediately with an error message mentioning `...Encoding not recognized...` your Windows OS most likely uses a rare UTF-8 based encoding.
```

Is it most likely or "most likely isn't using" since the encoding is not recognized? Change as necessary but see grammar change.
